src/cbmc_starter_kit/template-for-ci-workflow/proof_ci.yaml

```diff
@@ -0,0 +1,187 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+name: Run CBMC proofs
+on:
+  push:
+    branches-ignore:
+      - gh-pages
+
+# USAGE
+#
+# If you need to use different versions for tools like CBMC, modify this file:
+# .github/workflows/proof_ci_resources/config.yaml
+#
+# If you want the CI to use a different GitHub-hosted runner (which must still
+# be running Ubuntu 20.04), modify the value of this key:
+# jobs.run_cbmc_proofs.runs-on
+
+jobs:
+  run_cbmc_proofs:
+    runs-on: <__GITHUB_ACTIONS_RUNNER__>
+    name: run_cbmc_proofs
+    permissions:
+      contents: read
+      id-token: write
+      pull-requests: read
+    steps:
+      - name: Check out repository and submodules recursively
+        uses: actions/checkout@v3
+        with:
+          submodules: 'recursive'
+      - name: Parse config file
+        run: |
+          CONFIG_FILE='.github/workflows/proof_ci_resources/config.yaml'
+          cat << EOF >> $GITHUB_ENV
+          CADICAL_TAG=$(yq .cadical-tag $CONFIG_FILE)
+          CBMC_VERSION=$(yq .cbmc-version $CONFIG_FILE)
+          CBMC_VIEWER_VERSION=$(yq .cbmc-viewer-version $CONFIG_FILE)
+          KISSAT_TAG=$(yq .kissat-tag $CONFIG_FILE)
+          LITANI_VERSION=$(yq .litani-version $CONFIG_FILE)
+          PROOFS_DIR=$(yq .proofs-dir $CONFIG_FILE)
+          RUN_CBMC_PROOFS_COMMAND=$(yq .run-cbmc-proofs-command $CONFIG_FILE)
+          EOF
+      - name: Ensure CBMC, CBMC viewer, Litani versions have been specified
+        if: ${{ env.CBMC_VERSION == '' || env.CBMC_VIEWER_VERSION == '' || env.LITANI_VERSION == '' }}
+        shell: bash
+        run: |
+          should_exit=false
```

Comment:
I still don't think this variable, or the conditional on line 60, or any of the assignments to this variable, are necessary.

Reasoning:
* Because of the guard on line 44, we are guaranteed to enter at least one of the following `if` statements.
* Therefore, `should_exit` is guaranteed to be set true on all paths through this code. So it doesn't indicate anything, and the test on line 60 will always succeed.

This should also be clear from the fact that if the guard on line 44 is true, then that indicates an error, so we do need to exit unconditionally. There is no point in having a variable that tells us whether we need to exit, the fact that we run this step at all means that we do need to exit.

>> You could fix it by just writing exit 1 at the end rather than having the conditional.
>
> If you approach it the way that you suggest, and both CBMC_VERSION and LITANI_VERSION happen to be empty, then you will exit too early. With my approach, you see a statement being printed out for each of the tools that happen to be empty.

My suggestion is to write `exit 1` at the end of the step, not in each of the conditionals, so
```
if [ "${{ env.CBMC_VERSION }}" == "" ]; then
  echo "You must specify a CBMC version (e.g. 'latest' or '5.70.0')"
fi
if [ "${{ env.CBMC_VIEWER_VERSION }}" == "" ]; then
  echo "You must specify a CBMC viewer version (e.g. 'latest' or '3.6')"
fi
if [ "${{ env.LITANI_VERSION }}" == "" ]; then
  echo "You must specify a Litani version (e.g. 'latest' or '1.27.0')"
fi
exit 1
```
doing this will preserve the behavior you want, it will print all the errors before exiting.

>> so should_exit is vacuously true on all paths through this step.
>
> If should_exit was always true then this step would always exit with return code of 1, which is not the case based on what we have seen in past executions.

I believe it does always exit with 1, but the step doesn't always get executed in the first place because of the guard on line 44.

___

Here's an alternative suggestion. If you really want to keep `should_exit`, then just get rid of the guard on line 44. If you do that, I expect you'll now see the correct behavior because you fixed the conditional on line 60.

Reply:
> Here's an alternative suggestion. If you really want to keep should_exit, then just get rid of the guard on line 44. If you do that, I expect you'll now see the correct behavior because you fixed the conditional on line 60.

As discussed offline, I opted for this because if we ever decided to make tags for `kissat` and/or `cadical` mandatory, then one only needs to add if-statements in this _step_ instead of adding the if-statements *and* updating the `if` at for this GitHub step.